atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1831,6 +1831,11 @@ public boolean isUncommitted() {
         return state.get() == State.UNCOMMITTED;
     }
 
+    @Override
+    public boolean isDefinitivelyCommitted() {
```

Comment:
As I'm reading this code, I also realized we missed something here and it's why I'm kind of thinking a bit more broadly:

* in #commit, you don't check whether the method has been called, before doing #runPreCommitCallbacks. Meaning, technically if something throws there, someone could then call #commit again, and have them rerun. It's all WEIRD TERRITORY, but I think it's pointing at something deeper?

I feel like in general we're not modelling the state machine rigorously enough in all of this code. I would almost want to use internal state machine library to generate the code to make sure this is all done properly?